### ISSUE TITLE
False-positive E1121 when using dataclass with init=False

### ISSUE THREAD

**Issue by JohanKarlbergg** [pylint-dev/pylint#9389]
Similar issues https://github.com/pylint-dev/pylint/issues/4386 https://github.com/pylint-dev/pylint/issues/5857

### Bug description

```python
#!/usr/bin/python3
from dataclasses import dataclass
import inspect


@dataclass(init=False, frozen=True)
class A:
    a: str


@dataclass(frozen=True)
class B(A):
    b: str


if __name__ == "__main__":
    print(inspect.signature(B.__init__))
    print(B("a", "b"))
```

### Command used

```shell
pylint apa.py
```

### Pylint output

```python
************* Module apa
apa.py:18:10: E1121: Too many positional arguments for constructor call (too-many-function-args)
```

### Expected behavior

I don't expect an error

### Pylint version

```shell
pylint 3.0.3
astroid 3.0.2
Python 3.8.18 (default, Jan 18 2024, 10:10:55) 
[GCC 11.4.0]
```

### OS / Environment

PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
UBUNTU_CODENAME=jammy
